File: adventofcode2020/day19/solve.pl
my @rules;
$rules[$1] = $2 while (<> =~ /^(\d+): (.*)$/);

my @regexps;
sub regexp {
    my ($i) = @_;
    $regexps[$i] ||= $rules[$i] =~ /"(.*)"/ ? $1 :
        '(?:' . (join '|', map {join '', map {regexp($_)} split}
                 split / \| /, $rules[$i]) . ')';
}

my $rx = regexp(0);
my @messages = <>;
my $n = grep /^$rx$/, @messages;
print "Part 1: $n\n";

@regexps = (); # clear cache
regexp(42); regexp(31);
$regexps[8] = "$regexps[42]+";
$regexps[11] = "($regexps[42](?1)?$regexps[31])";
$rx = regexp(0);
$n = grep /^$rx$/, @messages;
print "regex: $rx";
print "Part 2: $n";

File: adventofcode2020/day19/solve.py
#!/bin/env python3
import sys
with open("input") as f:
    text = f.read()

rules_s, messages_s = text.split("\n\n")

rules = {}
for rs in rules_s.split("\n"):
    name, rd = rs.strip().split(": ")
    rules[name] = [x.split() for x in rd.split(" | ")]

messages = messages_s.split("\n")


def rule_to_regex(n):
    if rules[n][0][0][0] == '"':
        return rules[n][0][0][1]
    exp_o = []
    for option in rules[n]:
        subreg = []
        for x in option:
            if len(sys.argv) == 1:
                subreg.append(rule_to_regex(x))
                continue

            if x == "8":
                subreg.append(rule_to_regex("42")+"+")
            elif x == "11":
                subreg.append("("+rule_to_regex("42")+"(?1)?"+rule_to_regex("31")+")")
            else:
                subreg.append(rule_to_regex(x))
        exp_o.append("".join(subreg))
    if len(exp_o) == 1:
        return exp_o[0]
    return "(?:" + "|".join(exp_o) + ")"

rr = "^"+rule_to_regex('0')+"$"

valid_m = []
print(rr)
#print("I SHOULD HAVE DONE THIS IN PERL TO BEFGIN WITH")

